cura/CuraPackageManager.py

```diff
@@ -0,0 +1,357 @@
+                Logger.log("i", "Package [%s] version [%s] is newer than the installed version [%s], update it.",
+                           package_id, new_version, installed_version)
+                to_install_package = True
+
+        if to_install_package:
+            Logger.log("i", "Package [%s] version [%s] is scheduled to be installed.",
+                       package_id, package_info["package_version"])
+            # Copy the file to cache dir so we don't need to rely on the original file to be present
+            package_cache_dir = os.path.join(os.path.abspath(Resources.getCacheStoragePath()), "cura_packages")
+            if not os.path.exists(package_cache_dir):
+                os.makedirs(package_cache_dir, exist_ok=True)
+
+            target_file_path = os.path.join(package_cache_dir, package_id + ".curapackage")
+            shutil.copy2(filename, target_file_path)
+
+            self._to_install_package_dict[package_id] = {"package_info": package_info,
+                                                         "filename": target_file_path}
+            has_changes = True
+
+        self._saveManagementData()
+        if has_changes:
+            self.installedPackagesChanged.emit()
+
+    # Schedules the given package to be removed upon the next start.
+    @pyqtSlot(str)
+    def removePackage(self, package_id: str) -> None:
+        # Check the delayed installation and removal lists first
+        if not self.isPackageInstalled(package_id):
+            Logger.log("i", "Attempt to remove package [%s] that is not installed, do nothing.", package_id)
+            return
+
+        # Remove from the delayed installation list if present
+        if package_id in self._to_install_package_dict:
+            del self._to_install_package_dict[package_id]
+
+        # If the package has already been installed, schedule for a delayed removal
+        # if package_id in self._installed_package_dict:
+        #     self._to_remove_package_set.add(package_id)
+        # "Or rather don't because sometimes packages are not making it into the
+        # dict I guess." - Ian
+        self._to_remove_package_set.add(package_id)
+
+        self._saveManagementData()
+        self.installedPackagesChanged.emit()
+
+    # Removes everything associated with the given package ID.
+    def _purgePackage(self, package_id: str) -> None:
+        print("Purging",package_id)
+        # Get all folders that need to be checked for installed packages, including:
+        #  - materials
+        #  - qualities
+        #  - plugins
+        from cura.CuraApplication import CuraApplication
+        dirs_to_check = [
+            Resources.getStoragePath(CuraApplication.ResourceTypes.MaterialInstanceContainer),
+            Resources.getStoragePath(CuraApplication.ResourceTypes.QualityInstanceContainer),
+            os.path.join(os.path.abspath(Resources.getDataStoragePath()), "plugins"),
```

Comment:
This is missing the resource path to definitions. Maybe we should just check the data storage path?

Reply:
@LipuFei

Reply:
Now it's checking all subdirs in the data storage dir for package removal in 7ee085fe5